File: mobile/lib/candles.dart
// Ported from packages/backend/src/candles.ts. Generic OHLC candle bucketing:
// turns a raw (timestamp, close, volume) tick series into candles at a
// requested CEX-style interval. See candles_test.dart, ported from
// packages/backend/tests/candles.test.ts.

typedef ChartInterval = String; // "15m"|"1h"|"2h"|"4h"|"1d"|"2d"|"3d"|"1w"|"1M"

class Candle {
  final int timestamp;
  final double open;
  final double high;
  final double low;
  final double close;
  final double volume;

  Candle({
    required this.timestamp,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.volume,
  });
}

const int _minuteMs = 60000;
const int _hourMs = 60 * _minuteMs;

/// How much raw history to request from the source per interval -- see the
/// TS original for the full rationale (CoinGecko's free-tier 365-day cap on
/// historical `market_chart` queries, error_code 10012).
const Map<ChartInterval, int> intervalFetchDays = {
  "15m": 1,
  "1h": 7,
  "2h": 14,
  "4h": 30,
  "1d": 365,
  "2d": 365,
  "3d": 365,
  "1w": 365,
  "1M": 365,
};

int _floorToBucket(int ts, int bucketMs) => (ts ~/ bucketMs) * bucketMs;

String _dayKey(int ts) => DateTime.fromMillisecondsSinceEpoch(ts, isUtc: true).toIso8601String().substring(0, 10);

String _monthKey(int ts) => DateTime.fromMillisecondsSinceEpoch(ts, isUtc: true).toIso8601String().substring(0, 7);

DateTime _atDayOffset(DateTime t, int days) => DateTime.utc(t.year, t.month, t.day + days);

/// ISO week key (Monday-start, week 1 contains the year's first Thursday).
String isoWeekKey(int ts) {
  final d = DateTime.fromMillisecondsSinceEpoch(ts, isUtc: true);
  var target = DateTime.utc(d.year, d.month, d.day);
  final dayNr = target.weekday - 1; // Dart weekday: Mon=1..Sun=7 -> 0..6
  target = _atDayOffset(target, 3 - dayNr);
  var firstThursday = DateTime.utc(target.year, 1, 4);
  final firstThursdayDayNr = firstThursday.weekday - 1;
  firstThursday = _atDayOffset(firstThursday, 3 - firstThursdayDayNr);
  final week = 1 + (target.difference(firstThursday).inMilliseconds / (7 * 24 * 60 * 60 * 1000)).round();
  return "${target.year}-W${week.toString().padLeft(2, '0')}";
}

/// Groups raw ticks into candles by an arbitrary key -- the shared mechanics
/// behind every bucketing strategy. Assumes chronological input.
List<Candle> _groupTicksByKey(
  List<int> timestamps,
  List<double> closes,
  List<double> volumes,
  Object Function(int ts) keyFn,
) {
  final indicesByKey = <Object, List<int>>{};
  for (var i = 0; i < timestamps.length; i++) {
    final key = keyFn(timestamps[i]);
    (indicesByKey[key] ??= []).add(i);
  }

  return indicesByKey.values.map((indices) {
    final groupCloses = indices.map((i) => closes[i]).toList();
    final lastIndex = indices.last;
    return Candle(
      timestamp: timestamps[lastIndex],
      open: groupCloses.first,
      high: groupCloses.reduce((a, b) => a > b ? a : b),
      low: groupCloses.reduce((a, b) => a < b ? a : b),
      close: groupCloses.last,
      volume: volumes[lastIndex],
    );
  }).toList();
}

/// Regroups already-formed candles into coarser ones by calendar key.
List<Candle> _regroupCandles(List<Candle> candles, Object Function(int ts) keyFn) {
  final byKey = <Object, List<Candle>>{};
  for (final c in candles) {
    (byKey[keyFn(c.timestamp)] ??= []).add(c);
  }
  return byKey.values.map((group) {
    return Candle(
      timestamp: group.last.timestamp,
      open: group.first.open,
      high: group.map((c) => c.high).reduce((a, b) => a > b ? a : b),
      low: group.map((c) => c.low).reduce((a, b) => a < b ? a : b),
      close: group.last.close,
      volume: group.last.volume,
    );
  }).toList();
}

/// Chunks consecutive candles into groups of `size` -- for 2d/3d, which have
/// no natural calendar alignment, unlike week/month. Also reused directly by
/// the Binance klines path (connectors/binance_compatible.dart) for "2d",
/// the one interval Binance has no native granularity for.
List<Candle> chunkCandles(List<Candle> candles, int size) {
  final out = <Candle>[];
  for (var i = 0; i < candles.length; i += size) {
    final chunk = candles.sublist(i, (i + size).clamp(0, candles.length));
    out.add(Candle(
      timestamp: chunk.last.timestamp,
      open: chunk.first.open,
      high: chunk.map((c) => c.high).reduce((a, b) => a > b ? a : b),
      low: chunk.map((c) => c.low).reduce((a, b) => a < b ? a : b),
      close: chunk.last.close,
      volume: chunk.last.volume,
    ));
  }
  return out;
}

List<Candle> buildCandles(
  ChartInterval interval,
  List<int> timestamps,
  List<double> closes,
  List<double> volumes,
) {
  switch (interval) {
    case "15m":
      return _groupTicksByKey(timestamps, closes, volumes, (ts) => _floorToBucket(ts, 15 * _minuteMs));
    case "1h":
      return _groupTicksByKey(timestamps, closes, volumes, (ts) => _floorToBucket(ts, _hourMs));
    case "2h":
      return _groupTicksByKey(timestamps, closes, volumes, (ts) => _floorToBucket(ts, 2 * _hourMs));
    case "4h":
      return _groupTicksByKey(timestamps, closes, volumes, (ts) => _floorToBucket(ts, 4 * _hourMs));
    case "1d":
      return _groupTicksByKey(timestamps, closes, volumes, _dayKey);
    case "2d":
      return chunkCandles(_groupTicksByKey(timestamps, closes, volumes, _dayKey), 2);
    case "3d":
      return chunkCandles(_groupTicksByKey(timestamps, closes, volumes, _dayKey), 3);
    case "1w":
      return _regroupCandles(_groupTicksByKey(timestamps, closes, volumes, _dayKey), isoWeekKey);
    case "1M":
      return _regroupCandles(_groupTicksByKey(timestamps, closes, volumes, _dayKey), _monthKey);
    default:
      throw ArgumentError("Unhandled interval: $interval");
  }
}

File: packages/shared/src/candles.ts
// Generic OHLC candle bucketing: turns a raw (timestamp, close, volume) tick
// series into candles at a requested CEX-style interval. Pure and I/O-free so
// it's unit testable without live API access.
//
// Two bucketing strategies:
//  - sub-daily intervals (15m/1h/2h/4h): round each raw tick's timestamp down
//    to a fixed-width time bucket.
//  - daily+ intervals (1d/2d/3d/1w/1M): first collapse to one candle per UTC
//    day, then optionally regroup that daily series into 2/3-day chunks (no
//    natural calendar alignment, so just chunked) or calendar week/month
//    buckets (which do have a natural alignment).

export type ChartInterval = "15m" | "1h" | "2h" | "4h" | "1d" | "2d" | "3d" | "1w" | "1M";

export interface Candle {
  timestamp: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

const MINUTE_MS = 60_000;
const HOUR_MS = 60 * MINUTE_MS;
const DAY_MS = 24 * HOUR_MS;

/**
 * How much raw history to request from the source per interval. Sub-daily
 * intervals need finer source granularity and necessarily cover a shorter
 * span -- nobody expects 6 months of 15m candles, and CoinGecko's free tier
 * doesn't return intraday ticks beyond ~90 days anyway. Daily+ intervals
 * fetch a long span and derive coarser candles from daily bars.
 *
 * 365 is a hard ceiling here, not a stylistic choice: CoinGecko's free/Demo
 * tier rejects any `days` beyond 365 outright (error_code 10012, "Public API
 * users are limited to querying historical data within the past 365 days").
 * That means "1M" candles are capped at ~12 candles even for a token with
 * years of history -- well under what MACD/RSI need to warm up (~40) -- so
 * monthly analysis will usually report "not enough data" on this tier. That's
 * an honest reflection of the free tier's limit, not a bug; see the
 * dedicated message for it in routes/analysis.ts.
 */
export const INTERVAL_FETCH_DAYS: Record<ChartInterval, number> = {
  "15m": 1,
  "1h": 7,
  "2h": 14,
  "4h": 30,
  "1d": 365,
  "2d": 365,
  "3d": 365,
  "1w": 365,
  "1M": 365,
};

function floorToBucket(ts: number, bucketMs: number): number {
  return Math.floor(ts / bucketMs) * bucketMs;
}

function dayKey(ts: number): string {
  return new Date(ts).toISOString().slice(0, 10);
}

function monthKey(ts: number): string {
  return new Date(ts).toISOString().slice(0, 7);
}

/** ISO week key (Monday-start, week 1 contains the year's first Thursday). */
export function isoWeekKey(ts: number): string {
  const d = new Date(ts);
  const target = new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()));
  const dayNr = (target.getUTCDay() + 6) % 7;
  target.setUTCDate(target.getUTCDate() - dayNr + 3);
  const firstThursday = new Date(Date.UTC(target.getUTCFullYear(), 0, 4));
  const firstThursdayDayNr = (firstThursday.getUTCDay() + 6) % 7;
  firstThursday.setUTCDate(firstThursday.getUTCDate() - firstThursdayDayNr + 3);
  const week = 1 + Math.round((target.getTime() - firstThursday.getTime()) / (7 * DAY_MS));
  return `${target.getUTCFullYear()}-W${String(week).padStart(2, "0")}`;
}

/** Groups raw ticks into candles by an arbitrary string key -- the shared
 * mechanics behind every bucketing strategy. Assumes chronological input. */
function groupTicksByKey(
  timestamps: number[],
  closes: number[],
  volumes: number[],
  keyFn: (ts: number) => string
): Candle[] {
  const indicesByKey = new Map<string, number[]>();
  timestamps.forEach((ts, i) => {
    const key = keyFn(ts);
    if (!indicesByKey.has(key)) indicesByKey.set(key, []);
    indicesByKey.get(key)!.push(i);
  });

  return [...indicesByKey.values()].map((indices) => {
    const groupCloses = indices.map((i) => closes[i]);
    const lastIndex = indices[indices.length - 1];
    return {
      timestamp: timestamps[lastIndex],
      open: groupCloses[0],
      high: Math.max(...groupCloses),
      low: Math.min(...groupCloses),
      close: groupCloses[groupCloses.length - 1],
      volume: volumes[lastIndex],
    };
  });
}

/** Regroups already-formed candles into coarser ones by calendar key. */
function regroupCandles(candles: Candle[], keyFn: (ts: number) => string): Candle[] {
  const byKey = new Map<string, Candle[]>();
  for (const c of candles) {
    const key = keyFn(c.timestamp);
    if (!byKey.has(key)) byKey.set(key, []);
    byKey.get(key)!.push(c);
  }
  return [...byKey.values()].map((group) => ({
    timestamp: group[group.length - 1].timestamp,
    open: group[0].open,
    high: Math.max(...group.map((c) => c.high)),
    low: Math.min(...group.map((c) => c.low)),
    close: group[group.length - 1].close,
    volume: group[group.length - 1].volume,
  }));
}

/** Chunks consecutive candles into groups of `size` -- for 2d/3d, which have
 * no natural calendar alignment, unlike week/month. */
function chunkCandles(candles: Candle[], size: number): Candle[] {
  const out: Candle[] = [];
  for (let i = 0; i < candles.length; i += size) {
    const chunk = candles.slice(i, i + size);
    out.push({
      timestamp: chunk[chunk.length - 1].timestamp,
      open: chunk[0].open,
      high: Math.max(...chunk.map((c) => c.high)),
      low: Math.min(...chunk.map((c) => c.low)),
      close: chunk[chunk.length - 1].close,
      volume: chunk[chunk.length - 1].volume,
    });
  }
  return out;
}

export function buildCandles(
  interval: ChartInterval,
  timestamps: number[],
  closes: number[],
  volumes: number[]
): Candle[] {
  switch (interval) {
    case "15m":
      return groupTicksByKey(timestamps, closes, volumes, (ts) => String(floorToBucket(ts, 15 * MINUTE_MS)));
    case "1h":
      return groupTicksByKey(timestamps, closes, volumes, (ts) => String(floorToBucket(ts, HOUR_MS)));
    case "2h":
      return groupTicksByKey(timestamps, closes, volumes, (ts) => String(floorToBucket(ts, 2 * HOUR_MS)));
    case "4h":
      return groupTicksByKey(timestamps, closes, volumes, (ts) => String(floorToBucket(ts, 4 * HOUR_MS)));
    case "1d":
      return groupTicksByKey(timestamps, closes, volumes, dayKey);
    case "2d":
      return chunkCandles(groupTicksByKey(timestamps, closes, volumes, dayKey), 2);
    case "3d":
      return chunkCandles(groupTicksByKey(timestamps, closes, volumes, dayKey), 3);
    case "1w":
      return regroupCandles(groupTicksByKey(timestamps, closes, volumes, dayKey), isoWeekKey);
    case "1M":
      return regroupCandles(groupTicksByKey(timestamps, closes, volumes, dayKey), monthKey);
    default: {
      const exhaustive: never = interval;
      throw new Error(`Unhandled interval: ${exhaustive}`);
    }
  }
}
